apps/namadillo/src/atoms/integrations/services.ts

```diff
@@ -142,28 +142,28 @@ export const submitIbcTransfer = async (
 
 export const queryAndStoreRpc = async <T>(
   chain: Chain,
-  queryFn: QueryFn<T>,
-  rpcIndex = 0
+  queryFn: QueryFn<T>
 ): Promise<T> => {
   const { get, set } = getDefaultStore();
-  const workingRpcs = get(workingRpcsAtom);
-  const rpcAddress =
-    chain.chain_id in workingRpcs ?
-      workingRpcs[chain.chain_id]
-    : getRpcByIndex(chain, rpcIndex);
+  const lastRpc = get(rpcByChainAtom) || {};
+  const rpc =
+    chain.chain_id in lastRpc ?
+      lastRpc[chain.chain_id]
+    : getRpcByIndex(chain, 0);
 
   try {
-    const output = await queryFn(rpcAddress);
-    set(workingRpcsAtom, {
-      ...workingRpcs,
-      [chain.chain_id]: rpcAddress,
+    const output = await queryFn(rpc.address);
+    set(rpcByChainAtom, {
+      ...lastRpc,
+      [chain.chain_id]: { ...rpc },
     });
     return output;
   } catch (err) {
-    if (chain.chain_id in workingRpcs) {
-      delete workingRpcs[chain.chain_id];
-      set(workingRpcsAtom, { ...workingRpcs });
-    }
+    // On error, saves the next available rpc
+    set(rpcByChainAtom, {
+      ...lastRpc,
+      [chain.chain_id]: getRpcByIndex(chain, rpc.index + 1),
```

Comment:
Maybe in the future we could get from a random order. Loading all rpc urls into an atom and shuffling them. It could helps to don't have a bias for one or another rpc. But I see it is more work to do, so we can do it later

Reply:
We had that in the past, but we noticed that the first RPCs on the list were more reliable, so we switched to this solution